Source/WebCore/platform/graphics/texmap/TextureMapper.cpp

```diff
@@ -1320,6 +1320,68 @@ void TextureMapper::beginClip(const TransformationMatrix& modelViewMatrix, const
     clipStack().applyIfNeeded();
 }
 
+void TextureMapper::beginClip(const TransformationMatrix& modelViewMatrix, const FloatPolygon& polygon)
```

Comment:
It extends edges 1px.
https://github.com/WebKit/WebKit/blob/5da68fa7157872ef250bf4f159e14f4a2e3d58f1/Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp#L83

Reply:
I think it works. I need to check more closely the stencil usage. Most of this beginClip code is copy pasted from existing beginClip for FloatRect function

Reply:
This clipping actually clips antialiasing away so this needs to be fixed. We should render the stencil clip with same shader as layer rendering to get pixels perfect stencil clip for antialiased part. Unfortunately that shader is not compatible with how polygon clips need to be rendered and also intersection edge must not be antialiased so edge antialiasing needs to be selective. I'm currently evaluating couple of approaches to refactor antialiasing

Reply:
I don't mind disabling antialiasing in 3d rendering contexts if it's difficult.

Reply:
Currently, the antialiasing logic is designed specifically for rectangles, expecting that they fall within bounds 0..1 and are mapped to the correct size through the transformation matrix. Although the rectangle is divided into triangles, the expectation remains that these triangles stay within the rectangle shape. I’ve also observed cases where antialiasing interpolates inward, causing the edge to fade in an unintended direction.

Adapting the current logic to handle arbitrary polygons, which is needed for accurate clipping, seems quite challenging (if not impossible) as it stands. I’m implementing a new antialiasing shader that will work with arbitrary polygons but it will take an additional week or two to have it ready for review.

It might make sense to land this patch first and handle antialiasing in a separate patch, what do you think? The shader update could introduce layout test issues, so separating them probably makes it easier to review and maintain focus and clarity on correct issue.

Reply:
With this patch antialiasing doesn't completely fail in 3D. Non intersecting layers will have full antialiasing and also intersecting layers have some antialiasing but depending on transformation and perspective some of the antialiasing might be clipped off because stencil clip cannot use antialiasing shader logic.

Reply:
Sounds nice.